### PR TITLE
Add guideline for adding new algorithm

### PR DESCRIPTION
**Pull request by pan-x-c**
## Description

As the title says

## Checklist

Please check the following items before code is ready to be reviewed.

- [ ]  Code has passed all tests
- [ ]  Docstrings have been added/updated in Google Style
- [ ]  Documentation has been updated
- [ ]  Code is ready for review

### CODE REVIEW

File: docs/sphinx_doc/source/tutorial/trinity_programming_guide.md [hunk 1]
```diff
@@ -1,36 +1,44 @@
 # Developer Guide
 
-This guide introduces how to add new workflows to Trinity-RFT and provides relevant development guidelines.
+This guide introduces how to develop new modules to Trinity-RFT and provides relevant development guidelines.
```
Comment (hiyuchang) [inline docs/sphinx_doc/source/tutorial/trinity_programming_guide.md]:
```suggestion
This guide introduces how to develop new modules in Trinity-RFT and provides relevant development guidelines.
```

File: docs/sphinx_doc/source/tutorial/trinity_programming_guide.md [hunk 2]
```diff
@@ -314,6 +315,191 @@ trinity run --config <your_yaml_file>
...
+If you need to modify certain parameters, you can simply add the corresponding parameters within the `algorithm` section. For example, if you need to modify `repeat_times` and the initialization parameters of `AdvantageFn`, the modified `config.yaml` file would be as follows:
+
+```yaml
+# some other configs
+algorithm:
+  algorithm_type: "opmd"
+  repeat_times: 8
+  advantage_fn_args:
+    opmd_baseline: "logavgexp"
+    tau: 0.99
```
Comment (hiyuchang) [inline docs/sphinx_doc/source/tutorial/trinity_programming_guide.md:502]:
```suggestion
    tau: 0.99
  policy_loss_fn_args:
    tau: 0.99
```

File: docs/sphinx_doc/source/tutorial/trinity_programming_guide.md [hunk 3]
```diff
@@ -314,6 +315,196 @@ trinity run --config <your_yaml_file>
...
+- **Sample Strategy**: Responsible for reading experience data from the buffer module. By customizing this module, you can implement requirements like filtering experience data or mixed sampling from multiple data sources.
+- **Advantage Fn**: Responsible for calculating the Advantage and Returns of experience data.
+- **Policy Loss Fn**: Responsible for calculating the loss of the policy network.
+- **KL Fn**: Responsible for calculating KL Divergence, which is generally used in two places in existing RL algorithms: Reward Penalty and Actor Loss.
+- **Entropy Loss Fn**: Responsible for calculating the entropy loss of the policy network.
+
+We provide several implementations of above modules in `trinity/algorithm`.
+
+
+### Step 1: Implementing Algorithm Components
```
Comment (hiyuchang) [inline docs/sphinx_doc/source/tutorial/trinity_programming_guide.md]:
```suggestion
### Step 1: Implement Algorithm Components
```
Some titles use "doing"; some titles use "do". Try to keep consistent.

File: docs/sphinx_doc/source/tutorial/trinity_programming_guide.md [hunk 4]
```diff
@@ -1,36 +1,44 @@
 # Developer Guide
 
-This guide introduces how to add new workflows to Trinity-RFT and provides relevant development guidelines.
+This guide introduces how to develop new modules in Trinity-RFT and provides relevant development guidelines.
+
+Trinity-RFT consists of three main modules: **Explorer**, **Trainer** and **Buffer**.
+We decouple the RL pipeline into three modules to make it easier to customize and extend.
+Below is a table summarizing the modules that different types of developers need to focus on.
+
+| Developer Type | Focus Module | Key Component |
```
Comment (hiyuchang) [inline docs/sphinx_doc/source/tutorial/trinity_programming_guide.md]:
```suggestion
| Development Target | Core Module | Key Component |
```

File: docs/sphinx_doc/source/tutorial/trinity_programming_guide.md [hunk 5]
```diff
@@ -1,36 +1,44 @@
...
 
-
 - **Task** ({class}`trinity.common.workflows.Task`): Represents a data structure that can be converted into a `Workflow`. The content of the `Task` varies depending on the task type:
   - **Math problems**: A `Task` contains the problem description and the golden answer.
   - **Programming scenarios**: A `Task` includes the problem description, test cases, runtime environment, and other complex information.
 
-
-- **Workflow** ({class}`trinity.common.workflows.Workflow`): Can be understood as the running state of a `Task`. It defines the interaction flow between Agents and Environments, including logic similar to _Rollout_ and _Reward_ calculations in other frameworks. After execution, it generates a list of `Experience`. Trinity-RFT includes several built-in workflows:
+- **Workflow** ({class}`trinity.common.workflows.Workflow`): Can be understood as the running state of a `Task`. It defines the interaction flow between Agents and Environments, including logic similar to *Rollout* and *Reward* calculations in other frameworks. After execution, it generates a list of `Experience`. Trinity-RFT includes several built-in workflows:
```
Comment (hiyuchang) [inline docs/sphinx_doc/source/tutorial/trinity_programming_guide.md]: "Can be understood as the running state of a `Task`." I cannot understand this definition clearly. One suggestion is:

Describes how a `Task` is executed.

File: docs/sphinx_doc/source/tutorial/trinity_programming_guide.md [hunk 6]
```diff
@@ -40,12 +48,12 @@ Before starting development, it's important to understand several core concepts:
...
-  - **`workflow_args`** (`Dict`): A dictionary of parameters to facilitate the construction of `Workflow` instances. Provides more flexibility than `format_args` and `rollout_args` by using a dictionary. This field also comes from the YAML configuration file and can be set in `buffer.explorer_input.task_set.workflow_args`. Normally, you do not need to set this field.
+- **`workflow`** (`str`): The registered name of your workflow class. You can specify it in `buffer.explorer_input.taskset.default_workflow_type` of your YAML config file.
+- **`reward_fn`** (`Optional[str]`): The registered name of your reward function. You can specify it in `buffer.explorer_input.taskset.default_reward_fn_type`. Note that some workflows already include built-in reward calculation; in such cases, you can omit this field.
+- **`raw_task`** (`Dict`): An record of raw data in `Dict` format. For highly customized workflow, you can directly use `raw_task` to initialize your `Workflow` instance without relying on the following fields.
```
Comment (hiyuchang) [inline docs/sphinx_doc/source/tutorial/trinity_programming_guide.md]:
```suggestion
- **`raw_task`** (`Dict`): A record of raw data in `Dict` format. For highly customized workflow, you can directly use `raw_task` to initialize your `Workflow` instance without relying on the following fields.
```

File: docs/sphinx_doc/source/tutorial/trinity_programming_guide.md [hunk 7]
```diff
@@ -314,6 +315,196 @@ trinity run --config <your_yaml_file>
...
 ---
 
+## Algorithms (For RL Algorithm Developers)
+
+Trinity-RFT provides a standardized process for implementing new algorithms.
+
+### Step 0: Basic Concepts of Algorithm Module
+
+In Trinity-RFT, the algorithm module is primarily responsible for extracting experience data from the Replay Buffer during the RL process and calculating the loss to update models based on this data.
+To avoid implementing a new Trainer class each time a new algorithm is added, we have decomposed the representative PPO algorithm process into multiple sub-modules to adapt to various algorithms.
```
Comment (hiyuchang) [inline docs/sphinx_doc/source/tutorial/trinity_programming_guide.md:326]:
```suggestion
To avoid implementing a new Trainer class each time a new algorithm is added, we have decomposed the representative RL process into multiple sub-modules to adapt to various algorithms.
```

File: docs/sphinx_doc/source/tutorial/trinity_programming_guide.md [hunk 8]
```diff
@@ -314,6 +315,196 @@ trinity run --config <your_yaml_file>
...
+
+- **Sample Strategy**: Responsible for reading experience data from the buffer module. By customizing this module, you can implement requirements like filtering experience data or mixed sampling from multiple data sources.
```
Comment (hiyuchang) [inline docs/sphinx_doc/source/tutorial/trinity_programming_guide.md]:
```suggestion
- **Sample Strategy** ({class}`SampleStrategy`): Responsible for reading experience data from the buffer module. By customizing this module, you can implement requirements like filtering experience data or mixed sampling from multiple data sources.
```

File: docs/sphinx_doc/source/tutorial/trinity_programming_guide.md [hunk 9]
```diff
@@ -314,6 +315,196 @@ trinity run --config <your_yaml_file>
...
+- **Advantage Fn**: Responsible for calculating the Advantage and Returns of experience data.
+- **Policy Loss Fn**: Responsible for calculating the loss of the policy network.
+- **KL Fn**: Responsible for calculating KL Divergence, which is generally used in two places in existing RL algorithms: Reward Penalty and Actor Loss.
+- **Entropy Loss Fn**: Responsible for calculating the entropy loss of the policy network.
```
Comment (hiyuchang) [inline docs/sphinx_doc/source/tutorial/trinity_programming_guide.md]:
```suggestion
- **Advantage Function** ({class}`AdvantageFn`): Responsible for calculating the Advantage and Returns of experience data.
- **Policy Loss Function** ({class}`PolicyLossFn`): Responsible for calculating the loss of the policy network.
- **KL Loss Function** ({class}`KLFn`): Responsible for calculating KL Divergence, which is generally used in two places in existing RL algorithms: Reward Penalty and Actor Loss.
- **Entropy Loss Function** ({class}`PolicyLossFn`): Responsible for calculating the entropy loss of the policy network.
```

File: docs/sphinx_doc/source/tutorial/trinity_programming_guide.md [hunk 10]
```diff
@@ -314,6 +315,196 @@ trinity run --config <your_yaml_file>
...
+
+The main difference between OPMD and PPO algorithms lies in the calculation of Advantage and Policy Loss. Therefore, only new Advantage Fn and Policy Loss Fn modules need to be implemented.
+
+---
+
+#### Step 1.1: Implementing `AdvantageFn`
+
+Developers need to implement the {class}`trinity.algorithm.AdvantageFn` interface, which mainly includes two methods:
+
+- `__call__`: Calculates advantages and returns based on input experience data, records observable metrics during the calculation process, and returns the experience data containing advantages and returns as well as a metrics dictionary. The input experience data format is `verl`'s `DataProto`.
```
Comment (yanxi-chen) [inline docs/sphinx_doc/source/tutorial/trinity_programming_guide.md]:
```suggestion
- `__call__`: Calculates advantages and returns based on input experience data, records observable metrics during the calculation process, and returns the experience data containing advantages and returns as well as a metrics dictionary. The input experience data format is verl's `DataProto`.
```

File: docs/sphinx_doc/source/tutorial/trinity_programming_guide.md [hunk 11]
```diff
@@ -314,6 +315,196 @@ trinity run --config <your_yaml_file>
...
+The above steps implement the components needed for the algorithm, but these components are scattered and need to be configured in multiple places to take effect.
+
+To simplify configuration, Trinity-RFT provides {class}`trinity.algorithm.AlgorithmType` to describe a complete algorithm and registers it in {object}`trinity.algorithm.ALGORITHM_TYPE`, enabling one-click configuration.
+
+The `AlgorithmType` class includes the following attributes and methods:
+
+- `use_critic`: Whether to use the Critic model
+- `use_reference`: Whether to use the Reference model
+- `use_advantage`: Whether to calculate Advantage; if False, the `AdvantageFn` call will be skipped
+- `can_balance_batch`: Whether the algorithm can automatically balance batches
```
Comment (yanxi-chen) [inline docs/sphinx_doc/source/tutorial/trinity_programming_guide.md]:
```suggestion
- `can_balance_batch`: Whether the algorithm allows automatic balancing when splitting a batch into microbatches (which permute the order of samples)
```

File: docs/sphinx_doc/source/tutorial/trinity_programming_guide.md [hunk 12]
```diff
@@ -1,36 +1,44 @@
...
+
+Trinity-RFT consists of three main modules: **Explorer**, **Trainer** and **Buffer**.
+We decouple the RL pipeline into three modules to make it easier to customize and extend.
+Below is a table summarizing the modules that different types of developers need to focus on.
+
+| Developer Type | Focus Module | Key Component |
+|----------------|--------------|---------------|
+| Extend existing RL algorithms to new environments. | *Explorer* | `Workflow` |
+| Design new RL algorithms for higher model performance. | *Trainer* | `Algorithm` |
+| Enhance model performance from the data perspective. | *Buffer* | Data Processing Module (Coming soon) |
```
Comment (yanxi-chen) [inline docs/sphinx_doc/source/tutorial/trinity_programming_guide.md]:
```suggestion
| Apply existing RL algorithms to new environments. | *Explorer* | `Workflow` |
| Design new RL algorithms. | *Trainer* | `Algorithm` |
| Enhance the RL process from the data perspective. | *Buffer* | Data Processing Module (Coming soon) |
```

File: docs/sphinx_doc/source/tutorial/trinity_programming_guide.md (hunk 8 above)
Comment (yanxi-chen) [inline docs/sphinx_doc/source/tutorial/trinity_programming_guide.md]:
```suggestion
- **Sample Strategy**: Responsible for sampling experience data from the buffer module. By customizing this module, you can implement functionalities like filtering experience data or mixed sampling from multiple data sources.
```

Comment (yanxi-chen) [inline docs/sphinx_doc/source/tutorial/trinity_programming_guide.md]:
```suggestion
- **Sample Strategy**: Responsible for sampling experience data from the buffer module. By customizing this module, you can implement functionalities like filtering experience data or mixed sampling from multiple data sources.
```

File: docs/sphinx_doc/source/tutorial/trinity_programming_guide.md [hunk 13]
```diff
@@ -314,6 +315,196 @@ trinity run --config <your_yaml_file>
...
+Trinity-RFT provides a standardized process for implementing new algorithms.
+
+### Step 0: Basic Concepts of Algorithm Module
+
+In Trinity-RFT, the algorithm module is primarily responsible for extracting experience data from the Replay Buffer during the RL process and calculating the loss to update models based on this data.
+To avoid implementing a new Trainer class each time a new algorithm is added, we have decomposed the representative PPO algorithm process into multiple sub-modules to adapt to various algorithms.
+
+- **Sample Strategy**: Responsible for reading experience data from the buffer module. By customizing this module, you can implement requirements like filtering experience data or mixed sampling from multiple data sources.
+- **Advantage Fn**: Responsible for calculating the Advantage and Returns of experience data.
+- **Policy Loss Fn**: Responsible for calculating the loss of the policy network.
```
Comment (yanxi-chen) [inline docs/sphinx_doc/source/tutorial/trinity_programming_guide.md]:
```suggestion
- **Policy Loss Fn**: Responsible for calculating the core training loss of the policy network.
```
